src/lib.rs

```diff
@@ -323,87 +352,98 @@ impl AsyncHttpRangeReader {
             .parse()
             .map_err(|_err| AsyncHttpRangeReaderError::ContentLengthMissing)?;
 
-        // Allocate a memory map to hold the data
+        let builder = Self::default()
+            .url(head_response.url().clone())
+            .content_length(content_length as usize);
+
+        Ok(builder)
+    }
+
+    pub fn from_tail_response(tail_response: Response) -> Result<Self, AsyncHttpRangeReaderError> {
+        let content_range = ContentRange::parse(
+            tail_response
+                .headers()
+                .get(reqwest::header::CONTENT_RANGE)
+                .ok_or(AsyncHttpRangeReaderError::ContentRangeMissing)?
+                .to_str()
+                .map_err(|_err| AsyncHttpRangeReaderError::ContentRangeMissing)?,
+        );
+        let (start, finish, complete_length) = match content_range {
+            ContentRange::Bytes(ContentRangeBytes {
+                first_byte,
+                last_byte,
+                complete_length,
+            }) => (first_byte, last_byte, complete_length),
+            _ => return Err(AsyncHttpRangeReaderError::HttpRangeRequestUnsupported),
+        };
+
+        let requested_range =
+            SparseRange::from_range(complete_length - (finish - start)..complete_length);
+
+        // Configure the initial state of the streamer.
+        let mut streamer_state = StreamerState::default();
+        streamer_state
+            .requested_ranges
+            .push(complete_length - (finish - start)..complete_length);
+        let builder = Self::default()
+            .url(tail_response.url().clone())
+            .initial_tail_response(Some((tail_response, start)))
+            .requested_range(requested_range)
+            .content_length(complete_length as usize)
+            .streamer_state(streamer_state);
+        Ok(builder)
+    }
+
+    pub fn build(self) -> Result<AsyncHttpRangeReader, AsyncHttpRangeReaderBuilderError> {
+        let Some(client) = self.client else {
+            return Err(AsyncHttpRangeReaderBuilderError::MissingClient);
+        };
+        let Some(url) = self.url else {
+            return Err(AsyncHttpRangeReaderBuilderError::MissingUrl);
+        };
```

Comment:
I would prefer that if these fields are not optional, that a user provides them when constructing the builder. That turns this error from a runtime error into a compile time error.

Reply:
Hi, and thanks for the comments 👍🏼 Yeah, I see the `client` being required is probably good to have a compile-time check for.

Regarding `url` the reason I made it optional was that I was thinking if one calls `from_head_response`, it will set the url (so a caller who might want the existing behavior where you extract the url from the headers in the HEAD response could have that). On the other hand, somebody who might want to use a HEAD response (to validate that the headers confirm that Range requests are supported) **but** at the same time not use the header-url, can use the builder `url(...)` method to do so. WDYT?

Reply:
I started by making `client` required. Awaiting to hear your reaction to my comment about `url` before doing anything there 😊